pkg/logs/processor/processor.go

```diff
@@ -104,27 +129,89 @@ func (p *Processor) run() {
 
 	for {
 		select {
+		// Processing, usual main loop
+		// ---------------------------
+
 		case msg, ok := <-p.inputChan:
 			if !ok { // channel has been closed
 				return
 			}
-			p.processMessage(msg)
+
+			// if we have to wait for an SDS configuration to start processing & forwarding
+			// the logs, that's here that we buffer the message
+			if p.sds.buffering {
+				// buffer until we receive a configuration
+				p.sds.bufferMsg(msg)
+			} else {
+				// process the message
+				p.processMessage(msg)
+			}
+
 			p.mu.Lock() // block here if we're trying to flush synchronously
 			//nolint:staticcheck
 			p.mu.Unlock()
+
+		// SDS reconfiguration
+		// -------------------
+
 		case order := <-p.ReconfigChan:
 			p.mu.Lock()
-			if err := p.sds.Reconfigure(order); err != nil {
-				log.Errorf("Error while reconfiguring the SDS scanner: %v", err)
-				order.ResponseChan <- err
-			} else {
-				order.ResponseChan <- nil
-			}
+			p.applySDSReconfiguration(order)
 			p.mu.Unlock()
 		}
 	}
 }
 
+func (p *Processor) applySDSReconfiguration(order sds.ReconfigureOrder) {
+	isActive, err := p.sds.scanner.Reconfigure(order)
+	response := sds.ReconfigureResponse{
+		IsActive: isActive,
+		Err:      err,
+	}
+
+	if err != nil {
+		log.Errorf("Error while reconfiguring the SDS scanner: %v", err)
+	} else {
+		// no error while reconfiguring the SDS scanner and it is active?
+		// we should drain the buffered messages if any and stop the
+		// buffering mechanism.
+		if p.sds.buffering && isActive {
+			if p.sds.waitForConfig {
+				log.Debug("Processor ready with an SDS configuration.")
+				p.sds.buffering = false
+
+				// drain the buffer of messages if anything's in there
+				log.Info("SDS: draining", len(p.sds.buffer), "buffered messages")
+				for _, msg := range p.sds.buffer {
+					p.processMessage(msg)
+				}
+
+				p.sds.resetBuffer()
+			}
+		}
+		// no else case, the buffering is only a startup mechanism, after having
+		// enabled the SDS scanners, if they become inactive it is because the
+		// configuration has been sent like that.
+	}
+
+	order.ResponseChan <- response
+}
+
+func (s *sdsProcessor) bufferMsg(msg *message.Message) {
+	msgSize := len(msg.GetContent())
+	if msgSize+s.bufferedBytes < s.maxBufferSize {
+		s.buffer = append(s.buffer, msg)
+		s.bufferedBytes += msgSize
+	}
```

Comment:
Would it make sense to retain most recent messages, instead of the oldest?

Reply:
Agreed, as a user I think I'd find that preferable.

Reply:
I'll do that, I agree it's better but it'll make the buffering more complex (especially since we're not counting messages but bytes now)

Reply:
Now a FIFO